### ISSUE TITLE
The source mapping between the Xtend and the generated Java files are messed up

### ISSUE THREAD

**Issue by kittaakos** [eclipse/xtext-core#797]
I have experienced broken source mappings mainly in the LS related source when I was using xtext-core in Eclipse.

I am using `org.eclipse.xtext.ide_2.14.0.v20180424-1540.jar`.

Opening a type opens up the generated files instead of the Xtend file. It is not true for all types, but for a few. For instance, `TopologicalSorter` and `Document` are broken, but it is OK when opening the `DocumentExtensions` or `ServerModule`.

![topological_sort](https://user-images.githubusercontent.com/1405703/43130747-a4db4a20-8f37-11e8-885c-e91ffc21df10.gif)

Furthermore, the Eclipse debugger stops and opens the generated source instead of the Xtend one. Sometimes the lines are messed up in the Xtend files.

![screen shot 2018-07-24 at 11 46 08](https://user-images.githubusercontent.com/1405703/43130761-af57738e-8f37-11e8-95d2-f9fb6530071f.jpg)

I am happy to give additional input and/or verify something if it helps to resolve this issue. Thanks!

**Comment by cdietrich** [comment]
A PR fixing this would be welcome

**Comment by cdietrich** [comment]
@kittaakos btw your timestamp looks wired. are you sure you dont have a old nightly? that source and code versions match?

**Comment by cdietrich** [comment]
problem can be reproduced with

```
<project xmlns="http://maven.apache.org/POM/4.0.0"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<groupId>a</groupId>
	<artifactId>a</artifactId>
	<version>0.0.1-SNAPSHOT</version>

	<dependencies>
		<dependency>
			<artifactId>org.eclipse.xtext.ide</artifactId>
			<groupId>org.eclipse.xtext</groupId>
			<version>2.14.0</version>
		</dependency>
	</dependencies>
</project>
```
and org.eclipse.xtext.ide.server.Document

**Comment by cdietrich** [comment]
place to debug: org.eclipse.xtext.common.types.ui.trace.ClassFileBasedOpenerContributor.collectSourceFileOpeners(IEditorPart, IAcceptor<FileOpener>)

**Comment by cdietrich** [comment]
hmmm there are cases where it does not work at all

**Comment by szarnekow** [comment]
This is a known limitation of the eclipse APIs. Within the code that allows to override the editor for a given IJavaElement, we only get the simple name of the type. TopologicalSorter and Document do appear more often thus we do not find the real thing. In conclusion: We do not find the trace file. So we cannot navigate to the Xtend file.

**Comment by cdietrich** [comment]
But it’s not working for other files too. Still don’t know where the right place to debug is

**Comment by mark-christiaens** [comment]
I was wondering if the following is a manifestation of the same issue or is maybe a different issue?:

- Xtext 2.16
- I open `org.eclipse.xtext.ide.server.concurrent.RequestManager` (CTRL-SHIFT-T).  I don't have the actual project, `org.eclipse.xtext.ide` open in my workspace so Eclipse opens the associated source from the Jar-file.  
- On line 45, there is a reference to `ReadRequest`. 
- When I try to navigate on that line (F3) to `ReadRequest`, nothing happens

`ReadRequest` is a class present in the same source file.  In addition, there are no other `ReadRequest`s known to Eclipse that could confuse the issue.

**Comment by cdietrich** [comment]
this looks like the same issue if there is more than one ReadRequest

**Comment by cdietrich** [comment]
@mark-christiaens is your project a gradle or maven or pure eclipse project?

**Comment by cdietrich** [comment]
ander thing might be that cause its in the same file it might do nothing or dont do a select/reveal

**Comment by mark-christiaens** [comment]
I installed a complete new workspace using the Eclipse installer to make sure it has nothing to do with my own setup.  So this is a new workspace that contains the Xtext project itself and nothing else.  

![image](https://user-images.githubusercontent.com/813926/52212303-71b91c00-288c-11e9-9d44-e33d509b9594.png)

After the install, I (with the `org.eclipse.xtext.ide` project opened) can navigate to the `ReadRequest` from the actual Xtend source.  

Next, when I close the `org.eclipse.xtext.ide` project, and open the `RequestManager` I can no longer navigate to the `ReadRequest` mentioned in that `RequestManager`.  I can open the `ReadRequest` directly (CTRL-SHIFT-T) just fine.  I see the attached Xtend source just fine.

**Comment by mark-christiaens** [comment]
I doubt that it has anything to do with select/reveal since the navigation works as expected when I open the actual `org.eclipse.xtext.ide` project and work straight on the original sources.

**Comment by cdietrich** [comment]
org.eclipse.xtext.ui.editor.LanguageSpecificURIEditorOpener.open(URI, EReference, int, boolean) is a noop in this case. => different issue

**Comment by cdietrich** [comment]
@szarnekow does the code expect to find the .xtend files in the jar?

**Comment by szarnekow** [comment]
Looks like the thing is not even linked correctly, e.g. I don't get any hover info.

**Comment by mark-christiaens** [comment]
Indeed, I too see that "mark occurrences" works as intended.  So what's the conclusion?  Should I file a separate issue?

**Comment by szarnekow** [comment]
Nope, this issue here says it all. Initially I thought that no links are resolved at all, but that does not seem to be the case.
